logic/tool/TypeResolver.java

```diff
@@ -144,355 +84,175 @@ public Conjunction resolveLabels(Conjunction conjunction) {
         }
 
         private void convertIsa(TypeVariable owner, IsaConstraint isaConstraint) {
-            TypeVariable isaVar = convert(isaConstraint.type());
-            if (!isaConstraint.isExplicit()) owner.sub(isaConstraint.type(), false);
-            else if (isaConstraint.type().reference().isName()) owner.is(isaConstraint.type());
+            if (!isaConstraint.isExplicit())
+                resolverTraversal.sub(owner.id(), convert(isaConstraint.type()).id(), true);
+            else if (isaConstraint.type().reference().isName())
+                resolverTraversal.equalTypes(owner.id(), convert(isaConstraint.type()).id());
             else if (isaConstraint.type().label().isPresent())
-                owner.label(isaConstraint.type().label().get().properLabel());
-            else throw GraknException.of(ILLEGAL_STATE);
-            addNeighbours(owner, isaVar);
+                resolverTraversal.labels(owner.id(), isaConstraint.type().label().get().properLabel());
         }
 
-        private void convertIs(TypeVariable owner, grakn.core.pattern.constraint.thing.IsConstraint isConstraint) {
-            TypeVariable isVar = convert(isConstraint.variable());
-            owner.is(isVar);
-            addNeighbours(owner, isVar);
+        private void convertIs(TypeVariable owner, IsConstraint isConstraint) {
+            resolverTraversal.equalTypes(owner.id(), convert(isConstraint.variable()).id());
         }
 
-        private void convertIID(TypeVariable owner, IIDConstraint iidConstraint) {
-            owner.is(getTypeFromIID(iidConstraint.iid()));
-            TypeVariable iidVar = getTypeFromIID(iidConstraint.iid());
-            addNeighbours(owner, iidVar);
+        private void convertHas(TypeVariable owner, HasConstraint hasConstraint) {
+            resolverTraversal.owns(owner.id(), convert(hasConstraint.attribute()).id(), false);
         }
 
-        private TypeVariable getTypeFromIID(byte[] iid) {
-            assert conceptMgr.getThing(iid) != null;
-            Type type = conceptMgr.getThing(iid).getType();
-            return convert(retrieveTypeFromLabel(type.getLabel()));
+        private void convertRelation(TypeVariable owner, RelationConstraint constraint) {
+            for (RelationConstraint.RolePlayer rolePlayer : constraint.players()) {
+                TypeVariable playerResolver = convert(rolePlayer.player());
+                TypeVariable actingRoleResolver = convert(new TypeVariable(newSystemId()));
+                if (rolePlayer.roleType().isPresent()) {
+                    TypeVariable roleTypeResolver = convert(rolePlayer.roleType().get());
+                    resolverTraversal.sub(actingRoleResolver.id(), roleTypeResolver.id(), true);
+                }
+                resolverTraversal.relates(owner.id(), actingRoleResolver.id());
+                resolverTraversal.plays(playerResolver.id(), actingRoleResolver.id());
```

Comment:
When converting relation, we have to consider the possibility that the role type may be a super of what is actually being played by the role player. consider, so example:

```
define 
  partnership sub relation, relates partner;
  marriage sub partnership, relates spouse as partner;
  
  person sub entity, plays marriage:spouse
```

and

```
match ($role: $p) isa marriage;
```

clearly `marriage:spouse` is one possibility for foe `$role`, but so also is `parnternship:partner`. So we need one resolver for `$role` and another to link it to the particular role that the player (in this case `$p`) is playing.